### PR TITLE
Mulitstream guard rails

### PR DESCRIPTION
**Pull request by cbrochtrup**
If multistream raw value query is too large it will cripple the database. This adds in some guard rails to prevent users from taking down the system with a massive query.

- Add batched method in general utils
- Add test cases for the function
- Call the batching method on multistream queries
- Join resulting responses into overall table
- Remove some unnecessary list comprehensions

### CODE REVIEW

File: btrdb/utils/general.py [hunk 1]
```diff
@@ -36,7 +35,7 @@ def unpack_stream_descriptor(desc):
     return tags, anns
 
 
-def batched(iterable: Iterable[Any], n: int) -> Generator[Any, Any, Any]:
```
Comment (jleifnf) [inline btrdb/utils/general.py]: I think the Iterable would have been fine to use here if it was imported from `typing.Iterable`
but the Iterable [here](https://docs.python.org/3.8/library/typing.html#:~:text=Iterable%5BLoggedVar%5Bint%5D%5D)

Comment (jleifnf) [reply btrdb/utils/general.py]: i think the previous fail for walrus operator might also be from the typing hint was from using the collection's `Iterable`.

Comment (cbrochtrup) [reply btrdb/utils/general.py]: It was imported from typing `Iterable` before. I think the `Generator` isn't supported in earlier versions of Python and the variable is already named `iterable` so the partial type hint doesn't do much.

File: btrdb/stream.py [hunk 2]
```diff
@@ -2189,23 +2190,37 @@ def arrow_values(
...
-                    ],
-                )
-                data = pa.Table.from_arrays(
-                    [pa.array([]) for i in range(1 + len(self._streams))], schema=schema
-                )
+            data = None
+            # Batching the multistream queries is a stop gap to prevent
+            # the user from killing the DB by querying the raw values of
+            # too many streams at once.
+            for stream_batch in batched(self._streams, 20000):
```
Comment (jleifnf) [inline btrdb/stream.py:2199]: can we make a constant for this limit, `MAX_BATCH_STREAMS = 20000` or something similar to `INSERT_BATCH_SIZE`
```suggestion
            for stream_batch in batched(self._streams, MAX_BATCH_STREAMS):
```

**Review by jleifnf — CHANGES_REQUESTED** [review]
Might need a test for the actual stream batching behavior? Seems the arrow integrations are skipped in gh actions and might be more relevant of a test when we have coverage turned on to see if general arrow test case can satisfy it?
Thoughts, @justinGilmer ?

File: btrdb/stream.py [hunk 3]
```diff
@@ -2189,23 +2190,37 @@ def arrow_values(
...
-                    + [
-                        pa.field(str(s.uuid), pa.float64(), nullable=False)
-                        for s in self._streams
-                    ],
-                )
-                data = pa.Table.from_arrays(
-                    [pa.array([]) for i in range(1 + len(self._streams))], schema=schema
-                )
+            data = None
+            # Batching the multistream queries is a stop gap to prevent
```
[the remark this answers is not in the document]
Comment (cbrochtrup) [reply btrdb/stream.py:2197]: Good idea. I added one.

File: btrdb/stream.py (hunk 2 above)
Comment (cbrochtrup) [reply btrdb/stream.py]: Sounds good! Added.

**Comment by cbrochtrup** [comment]
Or, maybe it's pretty easy to turn those tests on? I don't know why they're skipped.

**Comment by cbrochtrup** [comment]
We've got server side guard rails now! No need for this 😄

**Pull request closed without merging**